### ISSUE TITLE
Sorting supported to extension fields?

### ISSUE THREAD

**Issue by Sameer2810** [microsoft/fhir-server#3616]
**Question**

Does sorting supported for extension fields

```[tasklist]
### Tasks
```

**Comment by brendankowitz** [comment]
Hi @Sameer2810. Under Health Data Services/SQL you can sort on string and date search parameters, so if you create a custom search parameter for the extension (then reindex) it would work.
Under ApiForFHIR/Cosmos, it supports creating the custom search parameter, but adding the capability to have it sortable would need a support ticket since that part isn't self-service.

**Comment by Sameer2810** [comment]
Hi @brendankowitz I have created Health Data Services and in that I have added ImagingStudy with date extension and perform reindex in that. The extension field (date field) is now searchable but for sorting it is not giving any result. 

Example - fhirurl/ImagingStudy?customdate={actual date}

For searching above is working 

Example - fhirurl/ImagingStudy?_sort=customdate  

Is above correct or I am missing something here

**Comment by EXPEkesheth** [comment]
@Sameer2810 - can you please help clarify if you are getting empty response or result with resources but not sorted?

**Comment by Sameer2810** [comment]
@EXPEkesheth - I am getting empty response

**Comment by mikaelweave** [comment]
@Sameer2810 - can you provide the expression for the `SearchParameter` resource you added for the custom search to drive the sorting?

**Comment by Sameer2810** [comment]
@mikaelweave following is the expression

ImagingStudy.extension.where(url = 'CreatedDate').value

**Comment by mikaelweave** [comment]
@Sameer2810 - The `url` for the extension is of type [uri](https://www.hl7.org/fhir/datatypes.html#uri) which must be in a certain format defined in [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986). Maybe this is causing the issue?

I'd suggest using a SearchParameter like below. I don't have time to test this now but this is what I'd try.

Also if there are other ways to capture this data without a custom extension I've found that best. For example, if ImagingStudy resources are always part of Procedure resources, you could maybe use `Procedure.recorded` and a chained search? This may not apply to your scenario but a good think to keep in mind.

```json
{
  "resourceType" : "SearchParameter",
  "id" : "resource-created-date",
  "url" : "http://mycompany.org/fhir/SearchParameter/resource-created-date",
  "version" : "1.0.0",
  "name" : "ResourceCreatedDate",
  "status" : "active",
  "date" : "2023-12-12",
  "publisher" : "My Company",
  "description" : "Returns the created date on resources.",
  "code" : "CreatedDate",
  "base" : [
    "Resource"
  ],
  "type" : "date",
  "expression" : "Resource.extension.where(url = 'http://mycompany.org/fhir/StructureDefinition/resource-created-date').extension.value.code"
}
```

**Comment by Sameer2810** [comment]
@mikaelweave  I did that what you mentioned, getting result but for some reason sorting not applied on result (for date type)

Also when I try with string type extension sorting applied to result (note - It gives result for those record which don't have that extension)

**Comment by Sameer2810** [comment]
@mikaelweave  and @EXPEkesheth Can you suggest further action

**Comment by mikaelweave** [comment]
@Sameer2810 - there was an error in the custom search parameter I posted. I updated the above comment and verified it works locally. Make sure the expression is correct and you remove the old search parameter first. Make sure to reindex and that all of your resources get reindexed by inspecting the response.

I find it easier to test on querying the data first vs sorting to ensure the custom search parameter is getting applied correctly (e.g. `https://{{hostname}}/Patient?CreatedDate=gt2020-01-1`).
